### DISCUSSION TITLE
Thinking with Portals

### DISCUSSION THREAD

**Discussion by glitchassassin** [glitchassassin/screeps-cartographer#33] [Ideas]
For Cartographer's next major milestone, I'd like to target portals.

This is only *partly* due to the fact that portals were used to eliminate me in SWC 212.

As I see it, the best approach is to replace findRoute with a custom implementation that treats portals like any other exit. We'll split the generated route on those portals, and then pathfind for each segment - from origin => portal and from the destination portal => target.

While we are replacing findRoute, this is a chance to deal with #20.

We can also take a look at another hurdle of efficient route finding - tracking whether a room is passable from exit A to exit B. This requires us to know the past three rooms in the path, rather than findRoute's two.

Any other considerations we ought to think about?

**Comment by rougeayrn** [comment]
Likely already considered, just not voiced: A built in way to cache seen portals would likely be beneficial given probability that vision would not be maintained. Love it, though! Can't wait to get to test it :)
